### PR TITLE
Better user interface

### PR DESCRIPTION
**Pull request by dvgica**
The existing `ArrivalsServer` and `ArrivalsGateway` interface was clunky. Users had to implement an abstract class with their routes, and had no control over the Akka HTTP server itself. If we look to Akka's built-in directives for a guide, we can see that directives and routes are usually in an `object` which gets imported like `import akka.http.scaladsl.server.Directives._`. Can we do the same?

Almost. The tricky bit is that our routes have dependencies like `HeaderAuthConfig`, `HttpProxy`, and `Metrics`. We can wrap up the common ones in an `ArrivalsContext` which gets passed implicitly to a route so that the routing DSL is still clean. However, `HeaderAuthConfig` presents some problems since we use it for path-dependent typing. Namely, we have types like `headerAuthConfig.AuthData` in the signatures of our routes, so `headerAuthConfig` needs to come in a previous parameter list. Of course, only the last parameter list can be implicit, so this would mean every route would have to explicitly pass a `HeaderAuthConfig`, which is really noisy. So the compromise is to make `trait AuthProxyController` into `class AuthProxyRoutes(headerAuthConfig: HeaderAuthConfig)`. This works for the path-dependent typing, and we can imitate the routes being in an `object` by doing:

``` scala
val routes = new AuthProxyRoutes(authConfig)
import routes._
```

Note: I believe once Scala 3 hits, we should be able to have arbitrary parameter lists be implicit, and then we can use `object AuthProxyRoutes`.

This cleanup works pretty well I think, you can see it in the `ExampleApp`. 

Other changes:
- Modified the example to not prefix everything with `/api` since it's not an API, and it's just noise
- Cleaned up the `AggregatorController` to not have so many type params by using `RunnableAggregator`
- Tightened up the path dependent types in the `*Routes` classes and was able to remove the casting.

### CODE REVIEW

File: arrivals-example/src/main/scala/com/pagerduty/arrivals/example/ExampleApp.scala [hunk 1]
```diff
@@ -43,14 +48,36 @@ object ExampleApp extends App {
...
+  val ExampleComposedResponseFilter = ExampleResponseFilterOne ~> ExampleResponseFilterTwo
+
+  val apiGatewayRoutes =
+    prefixProxyRoute("cats", CatsUpstream, ExampleComposedResponseFilter) ~
+      prefixAuthProxyRoute("dogs", DogsUpstream) ~
+      prefixAggregatorRoute("all", ExampleAggregator)
+
+  // start the API gateway with Arrivals routes
+  val apiGateway =
+    Await.result(Http().bindAndHandle(apiGatewayRoutes, "0.0.0.0", 8080), 5.seconds)
```
Comment (davidrusu) [inline arrivals-example/src/main/scala/com/pagerduty/arrivals/example/ExampleApp.scala:72]: 🤔 We are starting two `Http()` things here, one in `ArrivalsContext` up above and one here. What's the deal?

Can we replace this with:
```
Await.result(arrivalsCtx.http.bindAndHandle(apiGatewayRoutes, "0.0.0.0", 8080), 5.seconds)
```

File: arrivals/src/main/scala/com/pagerduty/arrivals/impl/aggregator/Aggregator.scala [hunk 2]
```diff
@@ -10,7 +10,8 @@ import com.pagerduty.arrivals.api.proxy.HttpProxy
 import scala.concurrent.{ExecutionContext, Future}
 
 trait Aggregator[AuthData, RequestKey, AccumulatedState, AddressingConfig]
-    extends api.aggregator.Aggregator[AuthData, RequestKey, AccumulatedState, AddressingConfig] {
+    extends api.aggregator.Aggregator[AuthData, RequestKey, AccumulatedState, AddressingConfig]
+    with RunnableAggregator[AuthData, AddressingConfig] {
 
   // implementation
   def apply(
```
Comment (davidrusu) [inline arrivals/src/main/scala/com/pagerduty/arrivals/impl/aggregator/Aggregator.scala]: `override def apply(...`  ?

File: arrivals/src/main/scala/com/pagerduty/arrivals/impl/ArrivalsContext.scala [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      override def executeWebSocketRequest[T](request: WebSocketRequest, clientFlow: Flow[Message, Message, T]) =
+        http.singleWebSocketRequest(request, clientFlow)(materializer)
+    }
+  }
+}
+
+case class ArrivalsContext[AddressingConfig](
+    addressingConfig: AddressingConfig,
+    entityConsumptionTimeout: FiniteDuration = 20.seconds,
+    buildHttpClient: (HttpExt, Materializer) => HttpClient = ArrivalsContext.buildHttpClient
```
Comment (davidrusu) [inline arrivals/src/main/scala/com/pagerduty/arrivals/impl/ArrivalsContext.scala:30]: 👍

**Review by davidrusu — CHANGES_REQUESTED** [review]

File: arrivals-example/src/main/scala/com/pagerduty/arrivals/example/ExampleApp.scala (hunk 1 above)
Comment (dvgica) [reply arrivals-example/src/main/scala/com/pagerduty/arrivals/example/ExampleApp.scala:72]: The `Http()` in `ArrivalsContext` is an internal implementation detail - it's the client used to proxy requests upstream. I think I would rather leave that as internal and not suggest that people use it to run their whole HTTP server.

File: arrivals/src/main/scala/com/pagerduty/arrivals/impl/aggregator/Aggregator.scala (hunk 2 above)
Comment (dvgica) [reply arrivals/src/main/scala/com/pagerduty/arrivals/impl/aggregator/Aggregator.scala]: Ah yes, good catch.

**Pull request merged**